### PR TITLE
Add Graph Transformer Layer (Dense Computation)

### PR DESCRIPTION
**Pull request by ZHITENGLI**
## Description

#4958 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes

Add Graph Transformer Layer (Dense Computation)

### CODE REVIEW

**Comment by rudongyu** [comment]
@dgl-bot

File: docs/source/api/python/nn-pytorch.rst [hunk 1]
```diff
@@ -118,6 +118,7 @@ Utility Modules
     ~dgl.nn.pytorch.graph_transformer.DegreeEncoder
     ~dgl.nn.pytorch.utils.LaplacianPosEnc
     ~dgl.nn.pytorch.graph_transformer.BiasedMultiheadAttention
+    ~dgl.nn.pytorch.graph_transformer.GraphTransformerLayer
```
Comment (mufeili) [inline docs/source/api/python/nn-pytorch.rst:121]: Is the indent here correct?

File: tests/pytorch/test_nn.py [hunk 2]
```diff
@@ -1801,3 +1801,22 @@ def test_BiasedMultiheadAttention(feat_size, num_heads, bias, attn_bias_type, at
...
 
     assert out.shape == (16, 100, feat_size)
+
+@pytest.mark.parametrize('feat_size', [128, 512])
+@pytest.mark.parametrize('ffn_size', [1024, 2048])
+@pytest.mark.parametrize('num_heads', [8, 16])
+@pytest.mark.parametrize('attn_bias_type', ['add', 'mul'])
+@pytest.mark.parametrize('norm_first', [True, False])
+@pytest.mark.parametrize('dropout', [0.1, 0.5])
+@pytest.mark.parametrize('activation', ['relu', 'leaky_relu'])
```
Comment (mufeili) [inline tests/pytorch/test_nn.py:1811]: L1805-1811 will result in 2^7 unit tests. It should be fine to test a single value for `feat_size`, `ffn_size`, `num_heads`, `dropout` and `activation`.

File: tests/pytorch/test_nn.py [hunk 3]
```diff
@@ -1801,3 +1801,22 @@ def test_BiasedMultiheadAttention(feat_size, num_heads, bias, attn_bias_type, at
...
+def test_GraphTransformerLayer(feat_size, ffn_size, num_heads, attn_bias_type,
+                               norm_first, dropout, activation):
```
Comment (mufeili) [inline tests/pytorch/test_nn.py:1813]: First declare
```
batch_size = 16
num_nodes = 100
```
to make the code more readable.

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 4]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+class GraphTransformerLayer(nn.Module):
+    r"""Graph Transformer Layer with Dense Multi-Head Attention, as introduced
+    in `Do Transformers Really Perform Bad for Graph Representation?
+    <https://arxiv.org/pdf/2106.05234>`__
+
+    Parameters
+    ----------
+    feat_size : int
+        Feature size.
+    ffn_size : int
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:237]: I prefer `hidden_size`.

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 5]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        Hidden size of feedforward layers.
+    num_heads : int
+        Number of attention heads, by which attr:`feat_size` is divisible.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:240]:
```suggestion
        Number of attention heads, by which :attr:`feat_size` is divisible.
```

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 6]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+    attn_bias_type : str, optional
+        The type of attention bias used for modifying attention. Selected from
+        'add' or 'mul'. Default: 'add'.
+
+        * 'add' is for additive attention bias.
+        * 'mul' is for multiplicative attention bias.
+    norm_first : bool, optional
+        If True, it performs layer norm before attention and feedforward
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:248]:
```suggestion
        If True, it performs layer normalization before attention and feedforward
```

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 7]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        operations. Otherwise, it applies layer norm after. Default: False.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:249]: "after" -> "afterwards"

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 8]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+    dropout : float, optional
+        Dropout probability. Default: 0.1.
+    activation : str, optional
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:252]: Why not use `callable` instead? e.g., users can pass `F.relu`

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 9]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        Activation function. Default: 'relu'.
+
+    Examples
+    --------
+    >>> import torch as th
+    >>> from dgl.nn import GraphTransformerLayer
+
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:259]: Better to declare variables first to make the code more readable

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 10]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        self.fc1 = nn.Linear(feat_size, ffn_size)
+        self.fc2 = nn.Linear(ffn_size, feat_size)
+
+        self.dropout = nn.Dropout(p=dropout)
+        self.activation_dropout = nn.Dropout(p=dropout)
+        self.activation = getattr(F, activation)
+        self.attn_layer_norm = nn.LayerNorm(feat_size)
+        self.ffn_layer_norm = nn.LayerNorm(feat_size)
+
+    def forward(self, ndata, attn_bias=None, attn_mask=None):
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:295]: Maybe use `feat` or `nfeat`, which has been widely used in the existing DGL modules

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 11]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        """Forward computation.
+
+        Parameters
+        ----------
+        ndata : torch.Tensor
+            A 3D input tensor. Shape: (batch_size, N, :attr:`feat_size`), where
+            N is the maximum number of nodes.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:302]: This seems to suggest the use of zero padding. Do we properly mask non-existing nodes in computations?

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 12]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
...
+        if not self.norm_first:
+            ndata = self.attn_layer_norm(ndata)
+
+        residual = ndata
+        if self.norm_first:
+            ndata = self.ffn_layer_norm(ndata)
+        ndata = self.activation(self.fc1(ndata))
+        ndata = self.activation_dropout(ndata)
+        ndata = self.fc2(ndata)
+        ndata = self.dropout(ndata)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/graph_transformer.py:330]: Consider using `nn.Sequential` for L327-330.

**Review by mufeili — CHANGES_REQUESTED** [review]
Some minor comments

File: docs/source/api/python/nn-pytorch.rst (hunk 1 above)
Comment (ZHITENGLI) [reply docs/source/api/python/nn-pytorch.rst:121]: `GraphTransformerLayer` is right according to dongyu's design doc, and `LaplacianPosEnc` will be adjusted by dongyu.

File: python/dgl/nn/pytorch/graph_transformer.py [hunk 13]
```diff
@@ -219,3 +223,113 @@ def forward(self, ndata, attn_bias=None, attn_mask=None):
         attn = self.out_proj(attn.reshape(N, bsz, self.feat_size).transpose(0, 1))
 
         return attn
+
+
+class GraphTransformerLayer(nn.Module):
+    r"""Graph Transformer Layer with Dense Multi-Head Attention, as introduced
+    in `Do Transformers Really Perform Bad for Graph Representation?
+    <https://arxiv.org/pdf/2106.05234>`__
```
[the remark this answers is not in the document]
Comment (ZHITENGLI) [reply python/dgl/nn/pytorch/graph_transformer.py:231]: Yes, it is the same as Graphormer, but other graph transformer models use almost the same structure except the dropout layers and we want to make this layer universal to them.

[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/nn/pytorch/graph_transformer.py:231]: This might need a bit further discussion. The existing DGL practice names modules based on the original model name, e.g. `GraphConv` and `SAGEConv` are similar to many GNN models. cc @jermainewang @BarclayII for discussion

File: python/dgl/nn/pytorch/graph_transformer.py (hunk 11 above)
Comment (ZHITENGLI) [reply python/dgl/nn/pytorch/graph_transformer.py:302]: It is achieved by `attn_mask`, but `attn_mask` can also be used to control the attention within the adjacency matrix besides masking nodes.

**Review by mufeili — APPROVED** [review]
I'm good other than the name of the module.

**Comment by mufeili** [comment]
@dgl-bot

**Comment by mufeili** [comment]
@dgl-bot

**Comment by mufeili** [comment]
@dgl-bot

**Comment by mufeili** [comment]
You need to resolve the conflicts.

**Comment by mufeili** [comment]
The file changes have got a bit messy. Did you change `PathEncoder`?

**Comment by ZHITENGLI** [comment]
> The file changes have got a bit messy. Did you change `PathEncoder`?

I just adjusted the order of different modules to make them more reasonable, did not change the content of `PathEncoder`.

**Comment by mufeili** [comment]
@dgl-bot
